<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Project: Analysis of vacancies from HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Preliminary data analysis

In [4]:
# at the beginning I want to look at the appearance and contents of the tables, so we select all the rows
# number of lines equals number of vacancies

query_1 = f'''select *
            from vacancies
         '''

In [5]:
vacancies_df = pd.read_sql_query(query_1, connection)
vacancies_df

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\3465271352.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_df = pd.read_sql_query(query_1, connection)


id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
49192  55589514           Специалист по информационным технологиям   
49193  55029668                     Руководитель процессного офиса   
49194  66172263                         Начальник сервисной службы   
49195  66180086                             Senior IOS разработчик   
49196  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
49192  Электронный документооборот\t1С: Предприятие 8...       Полный день   
49193  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
49194  Организаторские навыки\tУправление персоналом\...       Полный день   
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...       Полный день   
49196                                               None  Удаленная работа   

               experience           employment  salary_from  salary_to  \
0               Нет опыта     Полная занятость      64000.0        NaN   
1           От 3 до 6 лет     Полная занятость          NaN        NaN   
2           От 3 до 6 лет     Полная занятость          NaN        NaN   
3           От 3 до 6 лет     Полная занятость          NaN        NaN   
4               Нет опыта  Частичная занятость      15000.0        NaN   
...                   ...                  ...          ...        ...   
49192       От 3 до 6 лет     Полная занятость          NaN        NaN   
49193       От 3 до 6 лет     Полная занятость          NaN        NaN   
49194  От 1 года до 3 лет     Полная занятость      60000.0        NaN   
49195  От 1 года до 3 лет     Полная занятость     146456.0   244094.0   
49196       От 3 до 6 лет     Полная занятость          NaN        NaN   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
49192       66      5109332  
49193       26        51333  
49194     1930      4094700  
49195      160      4469939  
49196      160      4978627  

[49197 rows x 10 columns]

number of vacancies 49197

In [6]:
# Let's count the number of employers (table employers).
query_2 = f'''select * 
            from employers
         '''
# Similarly, look at the entire table, the number of employers is equal to the number of rows in the table

In [7]:
employers_df = pd.read_sql_query(query_2, connection)
employers_df

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\2513597085.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_df = pd.read_sql_query(query_2, connection)


id                                               name  area
0         2393                   Программный Продукт, ИТ-компания     1
1        72977                                          БАРС Груп    88
2         3155  Бест, Торгово-производственная компания, Екате...     3
3          675                                   КОРУС Консалтинг     2
4      1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88
...        ...                                                ...   ...
23496  1441774                                              МарКо    88
23497  4718575                                      MKomov Studio    73
23498  3702770                                  АКИБ Ипотека-банк  2759
23499  5636092                                        ГИМ-КЛИНИКА    99
23500  2064382                                             Рыббла   104

[23501 rows x 3 columns]

the number of employers is equal 23501

In [8]:
# Let's count the number of regions (areas table).
query_3 = f'''select *
             from areas
             '''

In [9]:
areas_df=pd.read_sql_query(query_3, connection)
areas_df

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\679926585.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_df=pd.read_sql_query(query_3, connection)


id              name
0     2758           Тбилиси
1        8            Майкоп
2     1180          Нерюнгри
3     1240       Новокузнецк
4        2   Санкт-Петербург
...    ...               ...
1357  5043  Звёздный Городок
1358  3934       Воротынск\n
1359  4825             Чемал
1360  3708         Усть-Нера
1361  2432         Каневская

[1362 rows x 2 columns]

the number of regions is equal 1362

In [10]:
# number of areas of activity in the database (industries table).
query_4 = f'''select *
           from industries
           '''

In [11]:
industries_df = pd.read_sql_query(query_4, connection)
industries_df

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\490431168.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_df = pd.read_sql_query(query_4, connection)


id                                               name
0      7.540                Разработка программного обеспечения
1      7.539  Системная интеграция,  автоматизации технологи...
2     27.550              Безалкогольные напитки (производство)
3     27.551  Безалкогольные напитки (продвижение, оптовая т...
4     13.664             Управление и эксплуатация недвижимости
..       ...                                                ...
289   37.626             Ассоциация в сфере культуры, искусства
290   42.603                              Табак  (производство)
291   29.532                           Рыболовство, рыбоводство
292   37.627                      Благотворительная организация
293  388.505  Оборудование для деревообработки, заготовки ле...

[294 rows x 2 columns]

the number of areas is equal 294

In [12]:
# We will make an additional request about employers who do not have posted vacancies
query_e = f'''select count(e.id) 
            from employers e
            left join vacancies v on e.id=v.employer_id
            where v.id is NULL 
         '''

In [13]:
emp_count_null = pd.read_sql_query(query_e, connection)
print('{} employers have no posted vacancies'.format(emp_count_null['count'][0]))

8595 employers have no posted vacancies


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\584767499.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_count_null = pd.read_sql_query(query_e, connection)


***

### *Сonclusions from preliminary data analysis*
Based on the requests, we see that there are about 50 thousand vacancies for 23.5 thousand employers - not so many, so we’ll make additional ones. request and see how many employers do not have active vacancies.

From additional query, we see that 8.5 thousand (about a third) employers from the database do not post vacancies. This may indicate that either many have closed or are posting vacancies under a different legal entity. face, enterprises have minimal staff turnover and there is no need for new employees. It may also be a sign of a decrease in the number of vacancies, job cuts. (It would be interesting to look at data for later periods here)

Upon request, we see that there are about 300 areas of activity, quite diverse. About 1.3 thousand regions are indicated, which is a lot, that is, vacancies are represented not only by large cities, but also by a lot of small settlements, and in the regions we also see countries and cities of the near abroad.

# Detailed analysis of vacancies

Let's make a request that will allow you to find out how many (cnt) vacancies there are in each region (area).
Let's sort by the number of vacancies in descending order.

In [14]:
query_5 = f'''select a.name area, count(v.id) cnt
           from vacancies v 
           join areas a on v.area_id=a.id
           group by area
           order by cnt desc
           '''

In [15]:
area_vac_df=pd.read_sql_query(query_5, connection)
area_vac_df.head(6)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\843855257.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  area_vac_df=pd.read_sql_query(query_5, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698

Let's make a request to determine how many vacancies have at least one of the two salary fields filled in.

In [16]:
query_6 = f'''select count(*)
           from vacancies
           where salary_from is not NULL or salary_to is not NULL
           '''

In [17]:
salary_fill=pd.read_sql_query(query_6, connection)
print('{} vacancies have salary fields filled in'.format(salary_fill['count'][0]))

24073 vacancies have salary fields filled in


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\3842905784.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_fill=pd.read_sql_query(query_6, connection)


Let's find the average values for the lower and upper limits of the salary range. Let's round the values to the nearest whole number.

In [18]:
query_7 = f'''select round(avg(salary_from),0) avg_from, round(avg(salary_to),0) avg_to
           from vacancies
           '''

In [19]:
salary_avg=pd.read_sql_query(query_7, connection)
salary_avg

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\453958802.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_avg=pd.read_sql_query(query_7, connection)


avg_from    avg_to
0   71065.0  110537.0

In [20]:
# Looking at the previous request, I would like to look at the average salary offered by industry.
# Therefore, we will make an additional request: we will display the name of the industry, the average salary 
# (look at the lower limit) and number of vacancies in the industry,
# we will remove industries where there are less than 10 vacancies
# sort by average salary in killing order

query_i = f'''select i.name industry, round(avg(v.salary_from),0) avg_from, count(v.id)
           from vacancies v
           join employers_industries ei on v.employer_id=ei.employer_id
           join industries i on ei.industry_id=i.id
           where (v.salary_from is not NULL)
           group by i.name
           having count(v.id)>=10
           order by 2 desc
                     '''

In [21]:
salary_avg_all=pd.read_sql_query(query_i, connection)
salary_avg_all.head(20)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\715869876.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_avg_all=pd.read_sql_query(query_i, connection)


industry  avg_from  count
0           Агрохимия (продвижение, оптовая торговля)  403077.0     13
1                            Агрохимия (производство)  338625.0     16
2                                      Игорный бизнес  147993.0     11
3                                          Зоомагазин  137419.0     31
4                   Корма для животных (производство)  136024.0     29
5   Бытовая химия, парфюмерия, косметика  (произво...  133807.0     70
6                                  Кадровые агентства  118131.0    575
7                      Учреждение соц.помощи и защиты  116786.0     14
8                           Ветеринарная деятельность  113171.0     13
9       Лакокрасочная продукция, сырье (производство)  106190.0     21
10  Управляющая, инвестиционная компания (управлен...  105287.0     71
11  Финансово-кредитное посредничество (биржа, бро...  100577.0    166
12  Бытовая химия, парфюмерия, косметика  (продвиж...  100576.0    131
13                               Снеки (производство)   98537.0     20
14  Рыба и морепродукты (продвижение, оптовая торг...   97813.0     16
15  Корма для животных (продвижение, оптовая торго...   97447.0     53
16  Лакокрасочная продукция, сырье (продвижение, о...   96662.0     27
17  Инфраструктурная компания в энергетике (СО ЕЭС...   95749.0    124
18                          Железнодорожные перевозки   93176.0     41
19           Консалтинг в области энергоэффективности   92218.0     15

Run a query that will return the number of vacancies for each combination of schedule type and employment type used in the vacancies. We sort the results in descending order of quantity.


In [22]:
query_8 = f'''select schedule, employment, count(*) 
           from vacancies
           group by 1,2
           order by 3 desc
           '''

In [23]:
sch_empl=pd.read_sql_query(query_8, connection)
sch_empl

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\530370018.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sch_empl=pd.read_sql_query(query_8, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

Let's execute a query that displays the values of the "experience" field in increasing order of the number of vacancies in which this experience option is indicated.

In [24]:
query_9 = f'''select experience
           from vacancies
           where experience is not NULL
           group by 1
           order by count(experience)
           '''

In [25]:
exp=pd.read_sql_query(query_9, connection)
exp

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\1468721607.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  exp=pd.read_sql_query(query_9, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

### *Conclusions from a detailed analysis of vacancies*

The query shows that the fewest vacancies are those with more than 6 years of experience. With what it can be connected? It can be assumed that specialists with extensive experience do not look for help so often, but more often in specialized communities (chat rooms). In general, a good experienced professional company will try to maintain self-confidence, so if an employee works for 3 to 6 years in the company and does not leave, then he is satisfied with everything, accordingly, requests for vacancies with such experience.
Most of the vacancies are in the category with work experience from 1 to 3 years, which is logical. The employee is already more or less experienced, requires less time for training and adaptation in the company, but is less demanding than an applicant with 3 or more years of experience.

From a request for the number of vacancies depending on the type of schedule and type of employment, it turns out that the most popular combination is Full Time/Full Time, which is logical. In second place - Remote work/Full-time, we can conclude that there are a large number of vacancies with this type of employment and most likely they will be shown. After all, remote work means the employer saves money on renting an office, equipment and some acceptable costs. (It is also interesting to look at data for past periods here).

From additional The request is obvious that the highest wages are made in wholesale trade. In most cases, high salary vacancies, most often among recruitment agencies, can be a recipe for disaster, because they provide vacancies in different entities.

From the additional request. The request states that half of the vacancies do not have salary fields filled in. Often the RFP field does not emerge if there are no or very few responses to the posted RFP and/or the employer wants to see which RFP is consistent in order to offer the optimal RFP and find a suitable employee (this may be for new positions in the company).

# Employer Analysis

Let's write a query that will allow you to find out which employers are in first and fifth place in terms of the number of vacancies.

In [26]:
query_10 = f'''(select e.name
           from vacancies v
           join employers e on e.id=v.employer_id
           group by 1
           order by count(v.id) desc
            limit 1)
           union all
           (select e.name
           from vacancies v
           join employers e on e.id=v.employer_id
           group by 1
           order by count(v.id) desc
           offset 4
           limit 1)
           '''

In [27]:
empl_vac=pd.read_sql_query(query_10, connection)
empl_vac

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\3644205195.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_vac=pd.read_sql_query(query_10, connection)


name
0         Яндекс
1  Газпром нефть

Let's write a query that for each region will display the number of employers and vacancies in it.
Among the regions in which there are no vacancies, we will find the one with the largest number of employers.

In [28]:
query_11 = f'''select a.name city, count(distinct e.id) count_employers, count(distinct v.id) count_vacancies
           from areas a
           left join employers e on a.id=e.area
           left join vacancies v on a.id=v.area_id
           group by city 
           having count(distinct v.id)=0
           order by 2 desc
           '''

In [29]:
empl_vac=pd.read_sql_query(query_11, connection)
empl_vac

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12008\3329030820.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_vac=pd.read_sql_query(query_11, connection)


city  count_employers  count_vacancies
0                            Россия              410                0
1                         Казахстан              207                0
2                Московская область               75                0
3                Краснодарский край               19                0
4                          Беларусь               18                0
..                              ...              ...              ...
588                        Кувандык                0                0
589                        Куйбышев                0                0
590  Куйбышево (Ростовская область)                0                0
591                        Култаево                0                0
592                        Кульсары                0                0

[593 rows x 3 columns]

For each employer, we count the number of regions in which it publishes its vacancies. Let's sort the results in descending order of quantity.

In [69]:
query_12 = f'''select e.name employers, count(distinct v.area_id) count_areas
           from employers e
           join vacancies v on v.employer_id=e.id
           group by employers
           order by count_areas desc
           '''

In [70]:
empl_areas=pd.read_sql_query(query_12, connection)
empl_areas

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\3363061791.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_areas=pd.read_sql_query(query_12, connection)


employers  count_areas
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
14761                  UniSol            1
14762            UNISTORY LLC            1
14763                   UNIT6            1
14764     United Distribution            1
14765              UNITEL LLC            1

[14766 rows x 2 columns]

In [34]:
# Let's make an additional request to see what kind of vacancies are posted by the employer that ranks 4th in terms of number
query_t = f'''select * 
           from employers e
           join vacancies v on v.employer_id=e.id
           where e.name='Поляков Денис Иванович'
          '''

In [35]:
q=pd.read_sql_query(query_t, connection)
q

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\1315612483.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  q=pd.read_sql_query(query_t, connection)


id                    name  area        id  \
0    5130287  Поляков Денис Иванович    22  55402339   
1    5130287  Поляков Денис Иванович    22  53316670   
2    5130287  Поляков Денис Иванович    22  54290525   
3    5130287  Поляков Денис Иванович    22  54303686   
4    5130287  Поляков Денис Иванович    22  66095025   
..       ...                     ...   ...       ...   
122  5130287  Поляков Денис Иванович    22  55186762   
123  5130287  Поляков Денис Иванович    22  55836122   
124  5130287  Поляков Денис Иванович    22  54534859   
125  5130287  Поляков Денис Иванович    22  55304208   
126  5130287  Поляков Денис Иванович    22  54427430   

                             name key_skills     schedule          experience  \
0             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
1             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
2             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
3             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
4             Компьютерный мастер       None  Полный день           Нет опыта   
..                            ...        ...          ...                 ...   
122           Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
123           Компьютерный мастер       None  Полный день           Нет опыта   
124           Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
125  Ученик компьютерного мастера       None  Полный день           Нет опыта   
126           Компьютерный мастер       None  Полный день  От 1 года до 3 лет   

           employment  salary_from  salary_to  area_id  employer_id  
0    Полная занятость        60000        NaN     1141      5130287  
1    Полная занятость        60000        NaN       47      5130287  
2    Полная занятость        60000        NaN       56      5130287  
3    Полная занятость        60000        NaN       70      5130287  
4    Полная занятость        40000    55000.0       20      5130287  
..                ...          ...        ...      ...          ...  
122  Полная занятость        60000        NaN     1243      5130287  
123  Полная занятость        50000        NaN       58      5130287  
124  Полная занятость        60000        NaN      212      5130287  
125        Стажировка        30000    50000.0        4      5130287  
126  Полная занятость        60000        NaN       25      5130287  

[127 rows x 13 columns]

Let's run a query to count the number of employers whose field of activity is not indicated.

In [36]:
query_13 = f'''select count(*)
           from employers e
           left join employers_industries ei on e.id=ei.employer_id
           where industry_id is NULL
           '''

In [37]:
empl_count=pd.read_sql_query(query_13, connection)
empl_count
print('{} employers do not indicate their field of activity'.format(empl_count['count'][0]))

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\1860687115.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_count=pd.read_sql_query(query_13, connection)


8419 employers do not indicate their field of activity


Let's write a request to find out the name of the company that is in third place in the alphabetical list (by name) of companies that have four areas of activity.

In [38]:
query_14 = f'''select e.name employers
           from employers e
           left join employers_industries ei on e.id=ei.employer_id
           group by employers
           having count(ei.employer_id)=4
           order by employers
           offset 2
           limit 1
           '''

In [39]:
empl_count2=pd.read_sql_query(query_14, connection)
print('{} is in third place among employers who have 4 areas of activity indicated'.format(empl_count2['employers'][0]))

2ГИС is in third place among employers who have 4 areas of activity indicated


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\2961383577.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_count2=pd.read_sql_query(query_14, connection)


Using a query, we will find out how many employers have “Software Development” listed as their field of activity.


In [40]:
query_15 = f'''select count(e.name) cnt
           from employers e
           join employers_industries ei on e.id=ei.employer_id
           join industries i on i.id=ei.industry_id
           where i.name='Разработка программного обеспечения'
           '''

In [41]:
empl_PO=pd.read_sql_query(query_15, connection)
print('{} employers indicated their field of activity as -"Разработка программного обеспечения"'.format(empl_PO['cnt'][0]))

3553 employers indicated their field of activity as -"Разработка программного обеспечения"


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\2696191071.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_PO=pd.read_sql_query(query_15, connection)


For «Яндекс» company, display a list of million-plus regions in which the company's vacancies are presented, along with the number of vacancies in these regions. Also add a Total line with the total number of vacancies for the company. Sort the results in ascending order of quantity.

Find a list of million-plus cities here: [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 



In [42]:
# code for getting a list of million-plus cities
!pip install beautifulsoup4 
from bs4 import BeautifulSoup

In [43]:
import requests
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) 
page = BeautifulSoup(response.text, 'html.parser')
lst = page.find_all('a')
lst_cities=tuple([link.text for link in lst[46:63]])
lst_cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград',
 '[7]')

In [44]:
query_16 = f'''(select a.name, count(v.id) cnt
           from vacancies v 
           join areas a on a.id=v.area_id
           join employers e on e.id=v.employer_id
           where e.name='Яндекс'
           group by a.name
           having a.name in ('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 '[a]',
 'Пермь',
 'Волгоград'))
 union all
 (select 'Total', count(*)
 from vacancies v 
           join areas a on a.id=v.area_id
           join employers e on e.id=v.employer_id
           where e.name='Яндекс' and a.name in ('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 '[a]',
 'Пермь',
 'Волгоград'))
  order by 2
           '''

In [45]:
ya=pd.read_sql_query(query_16, connection)
ya

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\3208637574.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ya=pd.read_sql_query(query_16, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

### *Conclusions from the analysis of employers*
We see that there are no vacancies if the employer's region is indicated in general - country, region, and there are also no vacancies in small cities. Perhaps such regions should not be indicated; a specific city would be better. If an applicant is looking for a vacancy, he is unlikely to search by region of Russia or the Moscow region, most likely he will look for a specific city.

A large number of employers (about a third) do not indicate the field of activity in the vacancy, which makes the search difficult for applicants, since many are looking for vacancies by choosing specific areas of activity. When posting a vacancy, it is worth recommending that employers indicate their field of activity, which will increase the number of views of the vacancy.

Completed additional a request to check the employer 'Polyakov Denis Ivanovich' (this is usually an individual entrepreneur), while posting vacancies in 88 regions, which is a lot for an individual entrepreneur (looks suspicious). From the request we see that these are the same vacancies “computer technician”. It may well be that such a network of workshops.

# Subject analysis

Let's check how many vacancies are related to the data?

We consider that a vacancy is related to data if its title contains the words 'data' or 'данн'.

In [46]:
query_17 = f'''select count(*)
           from vacancies
           where lower(name) like '%data%' or lower(name) like '%данн%'
           '''

In [47]:
cnt_data_vac=pd.read_sql_query(query_17, connection)
print('{} vacancies are related to data'.format(cnt_data_vac['count'][0]))

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\2692923845.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_data_vac=pd.read_sql_query(query_17, connection)


1771 vacancies are related to data


Let's count how many suitable vacancies there are for a beginning data scientist?
We will consider vacancies for data scientists to be those whose titles contain at least one of the following combinations:

 * 'data scientist'
 * 'data science'
 * 'исследователь данных'
 * 'ML' (there is no need to apply for vacancies here HTML)
 * 'machine learning'
 * 'машинн%обучен%'

**In the following tasks we will continue to work with vacancies under this condition*

We consider the following vacancies for Junior level specialists:
* there is a word in the title 'junior' *or*
* required experience - No experience(Нет опыта) *or*
* type of employment - Internship(Стажировка).
 

In [49]:
query_18 = f'''select count(name)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or 
           (name like '%ML%' and lower(name) not like '%html%'))
           and 
           (employment='Стажировка' or experience='Нет опыта' or lower(name) like '%junior%')
           '''

In [50]:
cnt_ds=pd.read_sql_query(query_18, connection)
print('{} - number of vacancies suitable for a beginning data scientist'.format(cnt_ds['count'][0]))

51 - number of vacancies suitable for a beginning data scientist


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\1638136091.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_ds=pd.read_sql_query(query_18, connection)


How many DS jobs are there that list SQL or postgres as a key skill?

** The criteria for classifying a vacancy as a DS are specified in the previous task.*

In [51]:
query_19 = f'''select count(*)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           and (lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%')
           '''

In [52]:
cnt_sk=pd.read_sql_query(query_19, connection)
print('{} - number of DS jobs that list SQL or postgres as a key skill'.format(cnt_sk['count'][0]))

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\1836153765.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_sk=pd.read_sql_query(query_19, connection)


201 - number of DS jobs that list SQL or postgres as a key skill


In [53]:
# we will make additional query and see how many vacancies there are that list SQL as a key skill
query_sql = f'''select count(*)
           from vacancies
           where lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%'
           '''

In [54]:
cnt_sql=pd.read_sql_query(query_sql, connection)
print('{} - total number of job openings that list SQL or postgres as a key skill'.format(cnt_sql['count'][0]))

10888 - total number of job openings that list SQL or postgres as a key skill


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\1839329056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_sql=pd.read_sql_query(query_sql, connection)


Let's check how popular Python is in employers' requirements for DS. To do this, calculate the number of job openings that list Python as a key skill.

In [55]:
query_20 = f'''select count(*)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or 
           (name like '%ML%'  and lower(name) not like '%html%'))
           and 
           (lower(key_skills) like '%python%')
           '''

In [56]:
cnt_pth=pd.read_sql_query(query_20, connection)
print('{} - number of DS jobs that list Python as a key skill'.format(cnt_pth['count'][0]))

351 - number of DS jobs that list Python as a key skill


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\443664897.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_pth=pd.read_sql_query(query_20, connection)


Let's calculate how many key skills are listed on average in DS vacancies?
We will round the answer to two decimal places after the separator point.

In [57]:
query_21 = f'''select avg(length(key_skills) - length(replace(key_skills,'\t',''))+1)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           '''

In [58]:
cnt_skills=pd.read_sql_query(query_21, connection)
cnt_skills

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\3407936094.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_skills=pd.read_sql_query(query_21, connection)


avg
0  6.406032

Let's write a query that allows us to calculate what the **average** salary for DS is for each type of experience required (a unique value from the *experience* field).

When solving the problem, we will take into account the following:
1. We consider only vacancies that have at least one of the two salary fields filled in.
2. If both salary fields are filled in, then we consider the salary for each vacancy as the sum of the two fields divided by 2. If only one of the fields is filled in, then we consider it as the salary for the vacancy.
3. If null is involved in the calculations, it will also result in null (see what the select 1 + null query returns). To avoid this situation, we will use the [coalesce] function.

Let’s find out what salary a data scientist with 3 to 6 years of experience can expect on average. Let's round the result to a whole number.

In [59]:
query_22 = f'''select experience, round(avg(COALESCE((salary_from+salary_to)/2, salary_from, salary_to)),0) avg_sal
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           and (salary_from is not NULL or salary_to is not NULL)
           group by experience
           '''

In [60]:
avg_sel=pd.read_sql_query(query_22, connection)
avg_sel

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\3017078929.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_sel=pd.read_sql_query(query_22, connection)


experience   avg_sal
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

In [61]:
# Let's make a request to see the average salaries for all vacancies
query_s = f'''select experience, round(avg(COALESCE((salary_from+salary_to)/2, salary_from, salary_to)),0) avg_sal
           from vacancies
           group by experience
           '''

In [62]:
avg_sel_all=pd.read_sql_query(query_s, connection)
avg_sel_all

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\4024140667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_sel_all=pd.read_sql_query(query_s, connection)


experience   avg_sal
0         Более 6 лет  166256.0
1       От 3 до 6 лет  133211.0
2           Нет опыта   40472.0
3  От 1 года до 3 лет   76541.0

In [63]:
# Let's make a request to see how many vacancies there are in DS
query_c = f'''select count(*)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           '''

In [64]:
count_vac_ds=pd.read_sql_query(query_c, connection)
count_vac_ds
print('{} - number of vacancies for DS'.format(count_vac_ds['count'][0]))

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_5704\168629528.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vac_ds=pd.read_sql_query(query_c, connection)


480 - number of vacancies for DS


***

### *Conclusions from subject analysis*
Please note in the request about the average salary depending on experience - there are no vacancies with more than 6 years of experience. The field of Data Science is quite new, especially for Russia, so people with more than 6 years of experience are rare and are unlikely to look for them on hh.

From the request for the average salary for DS vacancies depending on experience and additional. When asking about the average salary for all vacancies, we see that in DS salaries are almost 2 times higher than the average for all vacancies.

From the request we see that there are few vacancies for a beginning data scientist - 51 out of 480 vacancies (from the additional request) related to DS. We conclude that finding a job without experience in this area is quite difficult.

201 out of 480 job openings for DS list SQL as a key skill, which seems low, because in DS this query language is important. Therefore, we performed additional request to see how many vacancies contain SQL - 10888 - a lot. We conclude that it is important to know this query language. Many of these jobs are related to programming.

     P.S. I note that the requests were carried out based on the task criteria. But this request (concerning vacancies related to DS) could probably include extra vacancies (the 'ML' criterion), and many related to the 'DS' area do not fall into this area (for example, data analysts can be included here, etc.).

##  <center> *General conclusion about the project*

#### It is worth recommending that employers indicate their field of activity and indicate a specific city to increase views of the vacancy.
#### Many applicants, when searching for vacancies, filter by salary (from...), and in half of the vacancies the salary is not indicated; perhaps some employers should still indicate the salary range.

#### Predictions: vacancies with remote work will increase, knowledge of programming languages will be more in demand. As for trends, it would be good to compare with previous periods (estimate the number of vacancies and salaries for older periods).

#### We should consider how to improve the criteria to better count DS-related jobs.

#### You can conduct research on key skills, display them separately and look at those that are more in demand.